### PR TITLE
135 create modal for viewing a request

### PR DESCRIPTION
**Pull request by radu-constantin**
I've created the RequestModal component and the ResponseBox component.
Viewing requests and accepting responses work!
There is still a lot of work to be done and I've created separate issues for what I encountered.
I'm sure I didn't cover everything, so if you find anything else, go ahead and create an issue for it.

### CODE REVIEW

File: apps/frontend/src/components/CreateRequestModal/CreateRequestModal.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+          <label className={styles.label} htmlFor='content'>Content:</label>
+          <textarea className={styles.textarea} name='content' id='content' required></textarea>
+
+          <div className={styles.buttonContainer}>
+            <input className={styles.submit} type='submit' value='Submit' 
+            onClick={handleSubmit}></input>
+            <button
+              className={styles.button}
+              onClick={
```
Comment (nina-dk) [inline apps/frontend/src/components/CreateRequestModal/CreateRequestModal.tsx:33]: This closes the modal even if the form isn't successfully submitted, e.g. there's info missing.

File: apps/frontend/src/components/RequestBox/RequestBox.tsx [hunk 2]
```diff
@@ -46,18 +55,25 @@ export default function RequestBox({ requests }) {
...
-        <SearchFilterForm
-          filterStatus={requestsType}
-          setFilterStatus={setRequestsType}
-          requestSearchValue={requestSearchValue}
-          setRequestSearchValue={setRequestSearchValue}
-        />
+        <Form className={styles.form}>
+          <Form.Group>
+            <Form.Control type="text" placeholder="Search requests by title" value={searchCriteria.searchTerm} 
+            onChange={event => searchRequests((event.target).value)}></Form.Control>
```
Comment (nina-dk) [inline apps/frontend/src/components/RequestBox/RequestBox.tsx:61]: There's some bug in the searchRequests functionality. Sometimes it doesn't filter when I only input one character, and sometimes it filters but also keeps some requests that don't meet the criteria.

Comment (nina-dk) [reply apps/frontend/src/components/RequestBox/RequestBox.tsx:61]: found it. 😅 should do `startsWith` instead of `includes` unless that was the desired functionality to resemble keyword search and not an exact title match

File: apps/frontend/src/components/RequestBox/RequestBox.tsx [hunk 3]
```diff
@@ -46,18 +55,25 @@ export default function RequestBox({ requests }) {
...
+              <option value='ALL'>All Requests</option>
+            </Form.Select>
+          </div>
+        </Form>
       </div>
-      <RequestModal show={show} handleClose={handleClose} />
+      <CreateRequestModal show={show} handleClose={handleClose} />
       <div className={styles.container}>
-        {requestBoxes}
+        {requestBoxes.length !== 0 ? requestBoxes : 'Currently there are no requests that match your criteria!'}
```
Comment (nina-dk) [inline apps/frontend/src/components/RequestBox/RequestBox.tsx:75]: There's a styling bug here: the inputs overlap the section title and create request button when there are no results to display.

File: apps/frontend/src/components/RequestModal/RequestModal.tsx [hunk 4]
```diff
@@ -1,47 +1,87 @@
...
-        <Form method='post' className={styles.form} name='new-request-form'>
-          <label className={styles.label} htmlFor='title'>Title:</label>
-          <input className={styles.input} type='text' name='title' id='title' required minLength={4}></input>
+  async function handleCloseRequest() {
+    try {
+      setIsLoading(true);
+      await requestServices.closeRequest(String(request.id))
+      setIsLoading(false);
+      handleCloseModal();
+      revalidator.revalidate();
```
Comment (nina-dk) [inline apps/frontend/src/components/RequestModal/RequestModal.tsx:26]: Why is this necessary?

File: apps/frontend/src/components/RequestModal/RequestModal.tsx [hunk 5]
```diff
@@ -1,47 +1,87 @@
...
+  (OPEN or CLOSED) returns the corresponding JSX for the request modal.
+  */
+  function displayRequestActions(request: RequestType) {
+    let user = localStorage.getItem('user');
+    let username = user ? JSON.parse(user).username : null;
+    if (username === request.user.user_name && request.status === "OPEN") {
+      return (
+        <button className={`${styles.btn} ${styles.closeBtn}`} onClick={handleCloseRequest}>Close request</button>
+      )
+    } else if (!(username === request.user.user_name) && request.status === "OPEN") {
```
Comment (nina-dk) [inline apps/frontend/src/components/RequestModal/RequestModal.tsx:51]: could be `!==` but also, I think this first part of the condition is redundant

File: apps/backend/src/services/responseServices.ts [hunk 6]
```diff
@@ -129,6 +171,7 @@ const isUpdateResponseData = (obj: object): obj is UpdateResponseData => {
 const updateResponse = async (
   body: unknown,
   responseId: number,
+  userStatus: 'RESPONSE OWNER' | 'REQUEST OWNER',
```
Comment (nina-dk) [inline apps/backend/src/services/responseServices.ts:174]: we could create a type for this

**Review by nina-dk — CHANGES_REQUESTED** [review]
Excellent work as always, Radu. I've pinpointed a couple of issues.

**Comment by radu-constantin** [comment]
Does the app work for you after merging the changes on main? I'm getting a lot of errors on the frontend related to user_name

File: apps/frontend/src/components/CreateRequestModal/CreateRequestModal.tsx (hunk 1 above)
Comment (radu-constantin) [reply apps/frontend/src/components/CreateRequestModal/CreateRequestModal.tsx:33]: I need to create a separate issue out of it. We need to update the entire component, including its design.

File: apps/frontend/src/components/RequestBox/RequestBox.tsx (hunk 2 above)
Comment (radu-constantin) [reply apps/frontend/src/components/RequestBox/RequestBox.tsx:61]: Yes, I wanted to match based on keywords. I thought that maybe a user only wants to search for issues that have 'cat' in the title for example. What do you think?

File: apps/frontend/src/components/RequestBox/RequestBox.tsx (hunk 3 above)
Comment (radu-constantin) [reply apps/frontend/src/components/RequestBox/RequestBox.tsx:75]: Can you share a screenshot please?

File: apps/frontend/src/components/RequestModal/RequestModal.tsx (hunk 4 above)
Comment (radu-constantin) [reply apps/frontend/src/components/RequestModal/RequestModal.tsx:26]: The first line calls the backend to set the request as closed.
The second line signals that the loading is done so that the loading animation can stop.
The third line simply closes the modal.
The fourth line was added because I needed a way updated the list of Requests as soon as the backend was updated. 
For example, if you have the filter set to 'Open Requests' and you select a request and close it, I want the `Open Requests` list to be updated on the spot, so that the user won't see the request he just closed anymore. 
Normally, on FSO we were taught to achieve this by holding the request list within state and update that state list. I found this approach to be very cumbersome to achieve in this particular case. (It's been a while since I've done it so I can't explain why I found it so hard).
I decided in the end that it would be easier to just re-validate the request data by calling revalidator.revalidate(). https://reactrouter.com/en/main/hooks/use-revalidator
This re-loads the data and updates the component. It is possible that my approach is slower than simply updating state since it adds and extra Request-Response cycle.

File: apps/frontend/src/components/RequestModal/RequestModal.tsx (hunk 5 above)
Comment (radu-constantin) [reply apps/frontend/src/components/RequestModal/RequestModal.tsx]: Yes, you're right. It is redundant. Will fix!

File: apps/backend/src/services/responseServices.ts (hunk 6 above)
Comment (radu-constantin) [reply apps/backend/src/services/responseServices.ts:174]: Good ideea!

File: apps/frontend/src/components/RequestModal/RequestModal.tsx (hunk 4 above)
Comment (nina-dk) [reply apps/frontend/src/components/RequestModal/RequestModal.tsx:25]: Ok yeah this last line is what I was asking about, but using it as a last resort because useState was too complex makes sense. We could potentially look into it again in the future.

**Comment by nina-dk** [comment]
> Does the app work for you after merging the changes on main? I'm getting a lot of errors on the frontend related to user_name

For some reason, I had to fix those on my branch again. I'm guessing I did something wrong when merging. But it was nothing too difficult, just a couple of `user_name` accesses on the frontend instead of `username` which is the updated property name.

File: apps/frontend/src/components/RequestBox/RequestBox.tsx (hunk 2 above)
Comment (nina-dk) [reply apps/frontend/src/components/RequestBox/RequestBox.tsx:61]: agreed

File: apps/frontend/src/components/RequestBox/RequestBox.tsx (hunk 3 above)
Comment (nina-dk) [reply apps/frontend/src/components/RequestBox/RequestBox.tsx:75]: image

Comment (radu-constantin) [reply apps/frontend/src/components/RequestBox/RequestBox.tsx:75]: I can't manage to reproduce the bug  
Screenshot 2023-10-04 063425

Comment (nina-dk) [reply apps/frontend/src/components/RequestBox/RequestBox.tsx:75]: all right let's leave it for now since there will be more changes with the buttons and everything and I'll see if it persists in the future
